Reanalysis downloading module

ejsimley opened the pull request:
This pull request adds a module for downloading reanalysis data in the toolkits directory, called reanalysis_downloading. Currently, this module supports downloading MERRA-2 and ERA5 data using the PlanetOS API, but some documentation is provided explaining how to download data directly from NASA and Copernicus too. Documentation in the module explains how to obtain a PlanetOS API key, which is needed to use the module, as well. A method has also been added to the reanalysis data type class that allows loading data using the PlanetOS API (via the reanalysis_downloading toolkit module) in addition to loading from CSV files.

Within the reanalysis_downloading module, the main function used to download data is called "download_reanalysis_data_planetos". It returns a DataFrame containing the downloaded data and optionally saves the data as a csv file. It's arguments include: 

- the reanalysis product name ("merra2" or "era5")
- the lat/lon coordinates
- a few arguments used to specify the date range (see examples below)
- the atmospheric variables to download (defaults to the two wind speed components, temperature, and surface pressure)   
- the file and path name where a csv file containing the data should be saved (optional)
- the location of the file containing the API key

Note: before merging this we'll ask the PlanetOS team and NREL to review the documentation as well.

Here are some examples of using the function with the default atmospheric variables and without saving a CSV file, highlighting different ways of specifying the date range:

First, downloading the default date range, which is 20 years up to the end of the most recent full month:
```
from operational_analysis.toolkits import reanalysis_downloading as rd

lat, lon = (48.452, 5.588)
df = rd.download_reanalysis_data_planetos("era5",lat,lon)

df
                  datetime   u_100   v_100      t_2m     surf_pres
0      2001-08-01 00:00:00 -2.5625 -4.8125  291.7500  98484.851562
1      2001-08-01 01:00:00 -2.8750 -4.5000  290.8750  98490.906250
2      2001-08-01 02:00:00 -3.0625 -4.5625  290.3750  98477.046875
3      2001-08-01 03:00:00 -3.2500 -4.6875  289.7500  98478.187500
4      2001-08-01 04:00:00 -3.5625 -4.3750  289.5625  98478.132812
...                    ...     ...     ...       ...           ...
175315 2021-07-31 19:00:00  4.1250 -0.5000  290.0000  97289.796875
175316 2021-07-31 20:00:00  3.9375 -0.9375  289.1875  97285.867188
175317 2021-07-31 21:00:00  4.6875 -1.8750  287.8125  97288.343750
175318 2021-07-31 22:00:00  5.7500  1.1250  288.8125  97284.703125
175319 2021-07-31 23:00:00  5.9375 -1.8125  288.3125  97308.468750

[175320 rows x 5 columns]
```

Next, we'll download data for the date range from 2000-01-01 to 2005-07-15:
```
df = rd.download_reanalysis_data_planetos("era5",lat,lon,st_date="2000-01-01 00:00:00",en_date="2005-07-15 00:00:00")

df
                 datetime     u_100     v_100        t_2m     surf_pres
0     2000-01-01 00:00:00  2.410797  3.399582  275.588928  98387.429688
1     2000-01-01 01:00:00  2.662781  3.404541  275.622375  98394.093750
2     2000-01-01 02:00:00  2.945435  3.149887  275.503174  98405.484375
3     2000-01-01 03:00:00  3.265747  2.691406  275.703827  98422.265625
4     2000-01-01 04:00:00  3.489532  2.327530  275.806824  98435.046875
...                   ...       ...       ...         ...           ...
48524 2005-07-14 20:00:00 -3.000000 -1.000000  294.812500  97920.125000
48525 2005-07-14 21:00:00 -4.062500 -0.187500  293.375000  97928.125000
48526 2005-07-14 22:00:00 -2.687500  4.312500  292.500000  97936.250000
48527 2005-07-14 23:00:00 -1.375000  5.125000  292.500000  97923.625000
48528 2005-07-15 00:00:00 -0.187500  5.312500  292.875000  97903.750000

[48529 rows x 5 columns]
```

If we only specify the start date, it will download "num_years" of data starting on that date. Note, we're asking for MERRA-2 data now:
```
df = rd.download_reanalysis_data_planetos("merra2",lat,lon,st_date="2000-01-01 00:00:00",num_years=10)

df

                 datetime      u_50      v_50        t_2m     surf_pres
0     2000-01-01 00:30:00  2.752302  3.408454  273.226990  98509.031250
1     2000-01-01 01:30:00  2.949233  3.312825  273.238403  98532.351562
2     2000-01-01 02:30:00  3.327359  3.315226  273.190216  98559.171875
3     2000-01-01 03:30:00  3.849879  3.222857  273.234039  98568.304688
4     2000-01-01 04:30:00  3.964457  3.219257  273.700317  98577.156250
...                   ...       ...       ...         ...           ...
87667 2009-12-31 19:30:00  5.901794  1.654513  277.443542  95376.351562
87668 2009-12-31 20:30:00  5.220856  0.378700  276.895721  95362.507812
87669 2009-12-31 21:30:00  4.422727 -0.695482  276.098022  95343.648438
87670 2009-12-31 22:30:00  3.526426 -1.546500  275.594177  95334.453125
87671 2009-12-31 23:30:00  2.701854 -1.986580  275.162231  95315.953125

[87672 rows x 5 columns]
```

Similarly, if we only specify the end date, it will download "num_years" of data ending on that date.
```
df = rd.download_reanalysis_data_planetos("merra2",lat,lon,en_date="2000-01-01 00:00:00",num_years=10)

df
                 datetime      u_50      v_50        t_2m     surf_pres
0     1990-01-01 01:30:00  0.328658  2.941333  270.924133  97988.281250
1     1990-01-01 02:30:00  0.113014  2.962437  270.800140  97969.343750
2     1990-01-01 03:30:00 -0.067724  2.966995  270.768768  97947.398438
3     1990-01-01 04:30:00 -0.281913  2.952448  270.647156  97941.921875
4     1990-01-01 05:30:00 -0.480738  2.961390  270.590546  97939.265625
...                   ...       ...       ...         ...           ...
87643 1999-12-31 20:30:00  2.025297  3.811939  272.434448  98413.125000
87644 1999-12-31 21:30:00  2.199683  3.815449  272.300385  98435.867188
87645 1999-12-31 22:30:00  2.294813  3.794298  272.419586  98465.812500
87646 1999-12-31 23:30:00  2.517580  3.612560  272.907867  98490.921875
87647 2000-01-01 00:30:00  2.752302  3.408454  273.226990  98509.031250

[87648 rows x 5 columns]
```

Lastly, requesting 6 years, which by default will be the 6 years up to the end of the most recent full month:
```
df = rd.download_reanalysis_data_planetos("merra2",lat,lon,num_years=6)

df
                 datetime      u_50      v_50        t_2m     surf_pres
0     2015-08-01 00:30:00 -4.981873 -1.358198  286.007812  97589.085938
1     2015-08-01 01:30:00 -4.847740 -0.381847  285.551300  97576.781250
2     2015-08-01 02:30:00 -4.409152  0.083434  285.012390  97540.007812
3     2015-08-01 03:30:00 -3.736572 -0.024176  284.313782  97510.742188
4     2015-08-01 04:30:00 -2.728259  0.036082  284.508942  97545.765625
...                   ...       ...       ...         ...           ...
52603 2021-07-31 19:30:00  3.461423 -0.915823  288.748260  97289.398438
52604 2021-07-31 20:30:00  3.608768 -1.049836  288.247986  97263.187500
52605 2021-07-31 21:30:00  3.462683 -0.738081  287.779388  97301.031250
52606 2021-07-31 22:30:00  3.287869 -0.459484  287.502136  97329.554688
52607 2021-07-31 23:30:00  3.366580 -0.410493  287.400696  97313.023438

[52608 rows x 5 columns]
```

RHammond2 commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,316 @@
+    """
+
+    # Import PlanetOS apikey
+    if apikey_file is None:
```
The shortened version could be: 
```python
apikey = open(os.path.join(os.path.dirname(os.path.abspath(__file__)), apikey_file)).readlines()[0].strip()
```
or, if you use pathlib (another of my concerted efforts for making path objects easier to work with and read):
```python
from pathlib import Path  # this goes in the imports
apikey = open(Path(__file__).parent.resolve() / apikey_file).readlines()[0].strip()
```

ejsimley replied:
I wanted to allow the user to specify a different path to the file too, for example if they have the API key saved somewhere else for other PlanetOS purposes. This is why it only assumes the current path if apikey_file is None and otherwise assumes apikey_file includes the full path. I'm open to other suggestions for how to allow this flexibility, though.

RHammond2 replied:
I think that makes good sense so the user doesn't have to open the file, or include an API key in their workflow. Let's stick with what you did for both better usability and credential security.

RHammond2 commented on operational_analysis/types/reanalysis.py:
```diff
@@ -23,10 +25,53 @@ def __init__(self, engine="pandas"):
+                for product in self._products:
+                    self._product[product].load(path, "{}_{}".format(name, product))
+            elif format == "planetos":
+                for product in [p for p in self._products if p in ["merra2", "era5"]]:
```
If you wanted to be a bit slick here, you could write:
```python
for product in list(set(self._products) & set(("merra2", "era5")):
```

ejsimley replied:
Always looking for ways to be more slick! Changed to this style.

RHammond2 commented on operational_analysis/types/reanalysis.py:
```diff
@@ -23,10 +25,53 @@ def __init__(self, engine="pandas"):
+            elif format == "planetos":
+                for product in [p for p in self._products if p in ["merra2", "era5"]]:
+                    # Download from PlanetOS if csv file doesn't already exist
+                    if not os.path.exists(os.path.join(path, "{}_{}.csv".format(name, product))):
```
I know f-strings haven't quite caught on, but it does help to make a path string like this a little bit easier to read:
```python
if not os.path.exists(os.path.join(path, f"{name}_{product}.csv"))
```
and if you took it one step further with a [`pathlib.Path`](https://docs.python.org/3.8/library/pathlib.html#correspondence-to-tools-in-the-os-module) object:
```python
if not (Path(path) / f"{name}_{product}.csv").exists():
```

ejsimley replied:
Done

RHammond2 commented on operational_analysis/types/reanalysis.py:
```diff
@@ -23,10 +25,53 @@ def __init__(self, engine="pandas"):
+                            lat=lat,
+                            lon=lon,
+                            save_pathname=path,
+                            save_filename="{}_{}".format(name, product),
```
I'd push for using an f-string here to clarify what is being constructed: `save_filename=f"{name}_{product}"`

ejsimley replied:
Good suggestion. Updated this.

RHammond2 reviewed (commented):
Overall, I think this looks pretty good. Here are a few general comments I have:
 - The comments I made regarding proposed changes might be a bit more stylistic, so you don't have to take them, but could be helpful from a readability standpoint.
 - It might be helpful to have the internal functions for handling the request have a leading underscore to signal that this isn't really needed by the user, but this isn't totally necessary either.
 - I'd recommend adding a few tests for the helper functions to ensure that the variable dictionaries, start dates, end dates, etc. match what we're expecting for outputs.

nbodini commented:
From a scientific point of view this looks great, @ejsimley! I have no issues to report. 

The only discussion point I have is whether it makes more sense to add a function (or include this in the reanalysis_downloading) to calculate wind direction and air density instead of keeping  them as a necessary step to have in the project script. But totally a very minor point, I am fine either way!

ejsimley commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,316 @@
+    return st_date, en_date
+
+
+def get_st_end_dates_planetos(st_date, en_date, num_years, st_date_ds, en_date_ds):
```
This was just my convention, but I changed to "start_" and "end_" to make it easier to understand

ejsimley commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,316 @@
+    r = requests.get(url, params={"apikey": apikey})
+
+    st_date = pd.to_datetime(r.json()["subdatasets"][0]["temporalCoverage"]["start"] * 1e6)
+    en_date = pd.to_datetime(r.json()["subdatasets"][0]["temporalCoverage"]["end"] * 1e6)
```
changed to "start_date" and "end_date"

ejsimley commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,316 @@
+        3. If only st_date is defined, the final end date will be the lesser of num_years after the start date or the
+           last datetime of the dataset
+        4. If only en_date is defined, the final start date will be the greater of num_years before the end date or the
+           first datetime of the dataset
```
Thanks!

ejsimley commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,316 @@
+        st_date_new = st_date_ds
+
+        # If start date minute changes, update end date too
+        en_date_new = en_date_new.replace(minute=st_date_new.minute)
```
Adding num_years wouldn't necessarily work here because the user might have defined a specific end date they want to use. This would be a rare case, but the reason for this line is to make sure the difference between the start and end dates is an integer number of hours. Say the user specified start and end dates that are on the hour (e.g., 12:00). If the start date is out of bounds, and gets changed to the first date in the data set, which could be on a half hour (e.g., 00:30), then changing the end date's minute value to also be 30 minutes will make sure the PlanetOS request asks for the right number of total hours.

ejsimley commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,316 @@
+    return st_date_new, en_date_new
+
+
+def convert_resp_to_df(r, var_dict):
```
Good suggestion. I added a couple try/except blocks here.

ejsimley commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,316 @@
+
+    Args:
+        dataset (:obj:`string`): Dataset name ("merra2" or "era5")
+        lat (:obj:`float`): Latitude (degrees)
```
Yes, makes sense! Added this.

ejsimley commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,316 @@
+            temperature, and surface air pressure). Defaults to None.
+        save_pathname (:obj:`string`, optional): The path where the downloaded reanalysis data will be saved (if
+            defined). Defaults to None.
+        save_filename (:obj:`string`, optional): The file name used to save the downloaded reeanalysis data (if
```
fixed

ejsimley commented:
@RHammond2, @nbodini the latest commits should address your comments from back in August. I ended up revising quite a bit, so let me know if this looks good or if there are any more changes you'd like to make.

Rob: In addition to your inline comments, I added leading underscores to some of the internal/helper function in the reanalysis toolkit and made a unit test file for the reanalysis downloading toolkit to test a couple of the helper functions.     

Nicola: Based on your suggestion, I added the option to automatically derive wind speed and direction as well as air density from the available reanalysis variables. There's an option for this in the `reanalysis_downloading` toolkit function `download_reanalysis_data_planetos` and also a method to compute these variables in the `reanalysis` data type class, called `compute_derived_variables`.

I made another change based on some feedback from Patrick Duffy, who did some "beta testing" of the toolkit. Instead of requiring the user to provide a dictionary mapping the PlanetOS variable names to desired variable names, I added the argument `var_names` to `download_reanalysis_data_planetos` so users can just specify the PlanetOS variable names if they don't care about renaming them.

nbodini commented:
Thanks for your hard work on this, Eric, this looks great. 

Side note: I also got the same little fail in the test for check_simulation_results_gbm_daily while I was working on the AEP filter PR.. not sure what this was caused by, but I don't think it's related to your PR specifically.

RHammond2 commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,369 @@
+
+
+# Maps OpenOA-standard reanalysis product names to PlanetOS datasets
+dataset_names = {"merra2": "nasa_merra2_global", "era5": "ecmwf_era5_v2"}
```
I might have missed this before, but constants are typically in all caps in python. That said, we don't use constants much throughout OpenOA, so I could go either way on that convention. The main advantage would be being able to identify variables (constants) that sit outside the function more clearly, and signaling the intent of the variable (constant).

ejsimley replied:
I decided to change this dictionary constant to a function to add some string normalization and error handling.

RHammond2 commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,369 @@
+    # Download data from PlanetOS
+    url = "http://api.planetos.com/v1/datasets/" + dataset_names[dataset] + "/point?"
+
+    kwgs = {
+        "apikey": apikey,
+        "count": count,
+        "lon": lon,
+        "lat": lat,
+        "vars": ", ".join(var_names),
+        "time_start": start_date.strftime("%Y-%m-%dT%H:%M:%S"),
+        "time_end": end_date.strftime("%Y-%m-%dT%H:%M:%S"),
+    }
+
+    r = requests.get(url, params=kwgs)
```
Should we be allowing for the user to add in additional keyword arguments here by passing `**kwargs` into the function definition, or is that too niche of a use case?

ejsimley replied:
That's a good idea. I added a `**kwargs` argument that then gets appended to the standard kwargs dictionary used in the API call.

RHammond2 commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,369 @@
+        df["rho_kgm-3"] = met.compute_air_density(df["temperature_K"], df["surf_pres_Pa"])
+
+    if save_filename is not None:
+        df.to_csv(os.path.join(save_pathname, save_filename + ".csv"), index=False)
```
This comment is only because I noticed you used the `Path` object previously, and this is now the only use of the `os` library, so this could be amended to the following and we can remove an import from the top. That said, I see this is optional, and up to you. 
```python
df.to_csv(Path(save_pathname).resolve() / f"{save_filename}.csv", index=False)
```

ejsimley replied:
Good call, I made this change and removed `os` from the imports

RHammond2 commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,369 @@
+    Returns:
+        :obj:`dict`: Dictionary mapping reanalysis variable names from PlanetOS to standard OpenOA variable names
+    """
+
```
You might normalize the string input for this function to protect the user against typos by doing a `dataset = dataset.lower().strip()`

ejsimley replied:
Done. Also added error catching instead of returning None if the dataset is not merra2 or era5.

RHammond2 commented on test/unit/test_reanalysis_downloading_toolkit.py:
```diff
@@ -0,0 +1,284 @@
+from operational_analysis.toolkits import reanalysis_downloading as rd
+
+
+def test_default_var_dicts_planetos():
```
If you go with my comment in the function definition, I'd add a test for a typo like " MERRA2 ". Otherwise, I think if you add a case where it returns `None`, this test is good to go.

ejsimley replied:
Expanded the test cases and added a test for the new `_get_dataset_names` function too.

RHammond2 replied:
Looks good!

RHammond2 commented on test/unit/test_reanalysis_downloading_toolkit.py:
```diff
@@ -0,0 +1,284 @@
+import pandas as pd
+
+from operational_analysis.toolkits import reanalysis_downloading as rd
+
```
Very thorough, nice work!!

RHammond2 reviewed (commented):
I left a couple of small comments, and once they're resolved this is good to go.

ejsimley commented:
Hey @RHammond2, I addressed your last comments. Let me know if you think this looks good or if any of the additions should be modified.

jordanperr commented:
Triggering CI/CD pipeline

RHammond2 commented on operational_analysis/toolkits/reanalysis_downloading.py:
```diff
@@ -0,0 +1,389 @@
+    try:
+        return dataset_names[dataset]
+    except KeyError as e:
+        raise KeyError('Invalid dataset name. Currently, "merra2" and "era5" are supported.') from e
```
I like this control flow!

RHammond2 reviewed (approved):
This all looks good now, nice work!
